### PR TITLE
restrict cheats mutability

### PR DESCRIPTION
**Pull request by adhusson**
Many cheats should be view/pure. When you build tooling on top of cheats it's nice to be able to use regular view/pure modifiers.

edit: I haven't restricted the mutability of cheats that change the evm state, the environment, the file system, or the cheat interpreter state (eg `getRecordedLogs` stops recording so it's not `view`).

### CODE REVIEW

**Comment by mds1** [comment]
This seems useful, thanks @adhusson. I can see a use case for wanting internal or pure script methods that use a cheat code, and being forced to mark them as mutating due to cheat code use isn't great. 

First question for @gakonst @mattsse, is there any reason why all cheats are mutating? And is there any change this would require upstream?

Second question is what's the concrete distinction we're introducing? I think we're basically taking solidity's rules and generalizing them which seems reasonable:
- `pure` functions don't read state, they a solely a function of the input 
- `view` functions read and return some requested data, whether from the EVM, the filesystem, etc
- everything else mutates (e.g. `setEnv` or `warp`) or cannot have guarantees on what's touched (e.g. `ffi`)

If we agree on those definitions, a few things still aren't clear:
- Where does `expect*` and `record*` cheats fall? By those definitions they're view
- Does this mean `sign` and `addr` should be pure (currently they're view)

**Comment by mattsse** [comment]
> is there any reason why all cheats are mutating

I don't think so.

> I think we're basically taking solidity's rules and generalizing them which seems reasonable:

that sounds reasonable, and supportive of this.

> * Where does expect* and record* cheats fall? By those definitions they're view
> * Does this mean sign and addr should be pure (currently they're view)

I think we can make those changes for both cases.

**Comment by mds1** [comment]
> or the cheat interpreter state (eg `getRecordedLogs` stops recording so it's not view).

@adhusson it looks like you did make `getRecordedLogs` view, curious if that was a mistake or if you changed your mind on this definition?

Either way, I'd say lets:
- Add some comments on our pure/view/mutable definitions at the top of `Vm.sol` for future contributors (once we agree on the definitions)
- Make any required changes based on that definition, such as the changes in my two bullets above. There may be others, so worth doing a double check of all cheats
- Then merge!

**Comment by gakonst** [comment]
Sgtm

**Comment by adhusson** [comment]
> it looks like you did make `getRecordedLogs` view
@mds1 yes, it was a mistake! I've edited.

> Where does expect* and record* cheats fall? By those definitions they're view
They change the interpreter state as they modify the behaviour of later function calls, so I would say they're not view.

> Does this mean sign and addr should be pure (currently they're view)
They are pure!

I also changed `assume` to pure, I think that's the right thing to do formally speaking. Should be good to merge now.

**Review by mattsse — APPROVED** [review]
lgtm, 
good to merge @mds1 ?

**Comment by md1s** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

File: src/Vm.sol
Comment (mds1) [inline src/Vm.sol:10]:
```suggestion
// 2. Otherwise if you can be influenced by an earlier call, you're `view`,
// 3. Otherwise you're `pure`.
```

File: src/Vm.sol
Comment (mds1) [inline src/Vm.sol:72]:
```suggestion
    // Reads the entire content of file as binary. Path is relative to the project root. (path) => (data)
```

File: src/Vm.sol
Comment (mds1) [inline src/Vm.sol:9]: Would this be an accurate clarification? For example, `vm.projectRoot` reads some data but is not influenced by earlier calls

```suggestion
// 2. otherwise if you can be influenced by an earlier call, or if reading some state, you're `view`,
```

File: src/Vm.sol
Comment (mds1) [inline src/Vm.sol:148]: _(arbitrary comment location to start a thread)_

A few cheatcodes in `Vm` seem like they can be view too? I think just `isPersistent` and `activeFork`

**Review by mds1 — COMMENTED** [review]
sorry just a few small nits before merging!

**Comment by mds1** [comment]
@ZeroEkkusu would also love a sanity check from you on this PR if you don't mine

File: src/Vm.sol
Comment (adhusson) [reply src/Vm.sol]: Haha I thought about that but can't ffi change that value?

File: src/Vm.sol
Comment (adhusson) [reply src/Vm.sol:148]: Yes, will update!

File: src/Vm.sol
Comment (adhusson) [reply src/Vm.sol]: Nevermind, it can't! I'll include your edit.

**Review by mds1 — APPROVED** [review]
LGTM, thanks @adhusson! Will merge once @ZeroEkkusu takes a look

**Review by ZeroEkkusu — APPROVED** [review]
Sorry for the wait, LGTM if you settled on `record` and a few others being state-changing.

**Comment by mattsse** [comment]
@adhusson needs forge fmt

**Comment by mds1** [comment]
Thanks @adhusson!

**Pull request merged**
